### PR TITLE
Encapsulated error handling

### PR DESCRIPTION
**Pull request by mcollina**
This PR improves the error handling by allowing encapsulated error handler to call the upper-level error handler (or the fallback handler).

Ref https://github.com/fastify/fastify/issues/2860
Fixes https://github.com/fastify/fastify/issues/1436

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by mcollina** [comment]
@fastify/core I opened this one as draft to get early feedback. I would work to update the docs asap. Let me know what you think

---

@sergiz PTAL

File: lib/error-handler.js [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+  }
+}
+
+function handleError (reply, error, cb) {
+  console.log('handleError')
+  reply[kReplyIsRunningOnErrorHook] = false
+
+  const context = reply.context
+  if (reply[kReplyNextErrorHandler] === false) {
+    console.log('aaa')
```
Comment (kibertoad) [inline lib/error-handler.js:36]: is this intentional?

File: lib/error-handler.js [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    fallbackErrorHandler(error, reply, function (reply, payload) {
+      reply.raw.writeHead(reply.raw.statusCode, reply[kReplyHeaders])
+      reply.raw.end(payload)
+    })
+    return
+  }
+  const errorHandler = reply[kReplyNextErrorHandler] || context.errorHandler
+
+  // In case we the error handler throws, we set the next errorHandler so we can error again
```
Comment (kibertoad) [inline lib/error-handler.js:45]: "we" seems unnecessary

File: lib/error-handler.js [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  reply[kReplyNextErrorHandler] = Object.getPrototypeOf(errorHandler)
+
+  reply[kReplyHeaders]['content-length'] = undefined
+
+  const func = errorHandler.func
+  console.log(func)
```
Comment (kibertoad) [inline lib/error-handler.js:51]: is this intentional?

File: lib/error-handler.js [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    reply.code(500)
+    payload = serializeError({
+      error: statusCodes['500'],
+      message: err.message,
+      statusCode: 500
+    })
+  }
+
+  reply[kReplyHeaders]['content-type'] = 'application/json; charset=utf-8'
+  reply[kReplyHeaders]['content-length'] = '' + Buffer.byteLength(payload)
```
Comment (kibertoad) [inline lib/error-handler.js:123]: Is this faster than toString?

File: lib/error-handler.js (hunk 3 above)
Comment (mcollina) [reply lib/error-handler.js:51]: Yes. I'll remove all console.logs when I move it out of draft. Debugging this has been kind of hard and those statements helped me navigate.

File: test/reply-error.test.js [hunk 5]
```diff
@@ -622,3 +622,19 @@ test('should trigger error handlers if a sync route throws any non-error object'
...
+test('should trigger error handlers if a sync route throws undefined', async t => {
+  t.plan(1)
+
+  const fastify = Fastify()
+
+  const throwable = undefined
+
+  fastify.get('/', function async (req, reply) {
+    // eslint-disable-next-line no-throw-literal
+    throw throwable
```
Comment (kibertoad) [inline test/reply-error.test.js:635]: why not just throw undefined directly?

File: lib/error-handler.js (hunk 4 above)
Comment (mcollina) [reply lib/error-handler.js:123]: I copied and pasted this from elsewhere. Speed is not a concern here.

File: test/logger.test.js [hunk 6]
```diff
@@ -1369,7 +1370,7 @@ test('should not log the error if error handler is defined', t => {
         stream.once('data', line => {
           t.equal(line.level, 30, 'level is correct')
           t.equal(line.msg, 'request completed', 'message is set')
-          t.same(line.res, { statusCode: 500 }, 'status code is set')
+          t.same(line.res, { statusCode: 200 }, 'status code is set')
```
Comment (Eomm) [inline test/logger.test.js:1373]: Is no more set the status code to 500 by default?

Is the return code equals to 200 for this snippet?
```
fastify.setErrorHandler(function (error, request, reply) {
  this.log.error(error)
  reply.send({ success: false })
})
```

File: lib/reply.js [hunk 7]
```diff
@@ -508,7 +496,7 @@ function sendStream (payload, res, reply) {
 }
 
 function onErrorHook (reply, error, cb) {
-  if (reply.context.onError !== null && reply[kReplyErrorHandlerCalled] === true) {
+  if (reply.context.onError !== null && !reply[kReplyNextErrorHandler]) {
```
Comment (jsumners) [inline lib/reply.js:499]: Is the coercion check necessary? Or can it be `=== false`?

File: test/internals/handleRequest.test.js [hunk 8]
```diff
@@ -60,6 +55,7 @@ test('handler function - invalid schema', t => {
         }
       }
     },
+    errorHandler: { func: () => { t.pass('errorHandler called') } },
```
[the remark this answers is not in the document]
Comment (mcollina) [reply test/internals/handleRequest.test.js:58]: We need this complex object for the prototype inheritance chain, we can't have it with a simple function.

File: lib/route.js [hunk 9]
```diff
@@ -246,7 +247,7 @@ function buildRouting (options) {
 
       this.after((notHandledErr, done) => {
         // Send context async
-        context.errorHandler = opts.errorHandler || this[kErrorHandler]
+        context.errorHandler = opts.errorHandler ? buildErrorHandler(this[kErrorHandler], opts.errorHandler) : this[kErrorHandler]
```
[the remark this answers is not in the document]
Comment (mcollina) [reply lib/route.js:250]: It's not possible, we need the prototype inheritance chain.

File: test/logger.test.js (hunk 6 above)
Comment (mcollina) [reply test/logger.test.js:1373]: Yes, we changed it in https://github.com/fastify/fastify/pull/2609/files.

**Comment by mcollina** [comment]
This is now ready for review - I have just added docs.

File: docs/Errors.md [hunk 10]
```diff
@@ -37,7 +37,16 @@ To customize this behavior you should use [`setErrorHandler`](Server.md#seterror
 From the [Hooks documentation](Hooks.md#manage-errors-from-a-hook): 
 > If you get an error during the execution of your hook, just pass it to `done()` and Fastify will automatically close the request and send the appropriate error code to the user.
 
-If you have defined a custom error handler using `setErrorHandler` the error will be routed there, otherwise it will be routed to Fastify’s generic error handler. Fastify's generic error handler will use the header and status code in the Error object if it exists. The headers and status code will not be automatically set if a custom error handler is provided.
+If you have defined a custom error handler using `setErrorHandler` the error will
+be routed there. If `setErrorHandler` has been called multiple times, the error will be
+routed to the precedent error handler if the error handler or an hook `throw` or
+call `reply.send(error)`.
+Error handlers are fully encapsulated, so a `setErrorHandler` call in an encapsulated
+plugin will limit its effect there.
```
Comment (zekth) [inline docs/Errors.md:45]: Do we speak about `plugin` or `context` here? The encapsulation documentation speaks about `context`. Might be confusing

File: lib/route.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (Eomm) [reply lib/route.js:250]: Yeah, I have seen the trick right now with `Object.create() + Object.getPrototypeOf()`

So we should update the `Route.md` file accordingly

**Review by zekth — APPROVED** [review]
Just one comment on doc but LGTM.

good job on this!

File: lib/reply.js (hunk 7 above)
Comment (mcollina) [reply lib/reply.js:499]: The coercion is necessary.  There might be a bug hiding somewhere around this check as it was one of the most troubling spot while doing this change.

File: docs/Errors.md (hunk 10 above)
Comment (mcollina) [reply docs/Errors.md:45]: How would you rephrase? The current text sound clear to me but I'm definitely not the target audience for this text.

Comment (zekth) [reply docs/Errors.md:45]: I don't have a better phrasing to be honest. @jsumners might have an idea ?

File: lib/reply.js (hunk 7 above)
Comment (jsumners) [reply lib/reply.js:499]: Let's add a comment to that effect.

File: docs/Errors.md [hunk 11]
```diff
@@ -37,7 +37,16 @@ To customize this behavior you should use [`setErrorHandler`](Server.md#seterror
...
+The root error handler is Fastify’s generic error handler which will use the header
+and status code in the Error object if it exists.
+The headers and status code will not be automatically set if a custom error handler
+is provided.
```
Comment (jsumners) [inline docs/Errors.md:49]:
```suggestion
When a custom error handler has been defined through
[`setErrorHandler`](./Server.md#seterrorhandler), the custom error handler will
receive the error passed to the `done()` callback (or through other supported
automatic error handling mechanisms). If `setErrorHandler` has been used
multiple times to define multiple handlers, the error will be routed to the most
precedent handler defined within the error [encapsulation context](./Encapsulation.md).
Error handlers are fully encapsulated, so a `setErrorHandler` call within a
plugin will limit the error handler to that plugin's context.

The root error handler is Fastify's generic error handler. This error handler
will use the headers and status code in the `Error` object, if they exist. The
headers and status code will not be automatically set if a custom error handler
is provided.
```

File: docs/Errors.md (hunk 10 above)
Comment (jsumners) [reply docs/Errors.md:45]: See my suggestion below.

File: test/encapsulated-error-handler.test.js [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+const Fastify = require('..')
+
+test('encapuslates an error handler', async t => {
+  t.plan(3)
+
+  const fastify = Fastify()
+  fastify.register(async function (fastify) {
+    fastify.setErrorHandler(async function a (err) {
+      t.equal(err.message, 'kaboom')
+      throw new Error('catched')
```
Comment (sergiz) [inline test/encapsulated-error-handler.test.js:13]: It should be "caught".

File: lib/route.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (mcollina) [reply lib/route.js:250]: Why? Do we document `context.errorHandler`?

File: test/encapsulated-error-handler.test.js (hunk 12 above)
Comment (sergiz) [reply test/encapsulated-error-handler.test.js:13]: @mcollina You haven't changed all occurances. There are still two incorrect verbs used.

File: lib/route.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (Eomm) [reply lib/route.js:250]: No, I mean here: https://github.com/fastify/fastify/blob/main/docs/Routes.md#options
there is the `errorHandler` that is an object now and no more a function

[the remark this answers is not in the document]
Comment (mcollina) [reply lib/route.js:250]: That did not change, it's still a function. We wrap it internally.

**Review by Eomm — APPROVED** [review]

**Review by fox1t — APPROVED** [review]
Amazing feature!

**Comment by sergiz** [comment]
The first message incorrectly states that this PR fixes https://github.com/fastify/fastify/issues/2860. It doesn't make any changes in that regard.

**Comment by mcollina** [comment]
> The first message incorrectly states that this PR fixes #2860. It doesn't make any changes in that regard.

My reading of what is left of that issue is that throwing within error handlers was something not supported before.
This makes it the default mode of operation for calling the parent error handler.

Maybe I misunderstood that issue.

**Comment by sergiz** [comment]
The inconsistencies described there are in regards to throwing errors in connection with whether handler/error handler are sync or async. This PR doesn't make any changes in that regard. I will open a PR which addresses these things during the weekend.

**Comment by mcollina** [comment]
Ok!

Landing this now so you could use as a basis and not conflict.

**Review by delvedor — APPROVED** [review]
A delayed LGTM!

**Pull request merged**